### PR TITLE
[master] Fix detection of Salt codename by "salt_version" execution module.

### LINKED ISSUE #64306: [BUG] modules.salt_version appears broken in Sulfur
**Description**
The salt_version module appears to be broken. Per this page: https://docs.saltproject.io/en/latest/topics/releases/version_numbers.html#version-numbers, Salt 3006.1 should have the codename `Sulfur`. However:

```
salt --versions-report
Salt Version:
              Salt: 3006.1
...

salt salt01 grains.get saltversion
salt01:
    3006.1

salt salt01 grains.get saltversioninfo
salt01:
    - 3006
    - 1
```

```
salt salt01 salt_version.get_release_number 'Sulfur'
salt01:
    No version assigned.

salt salt01 salt_version.equal 'Sulfur'
salt01:
    False

salt salt01 salt_version.less_than 'Sulfur'
salt01:
    True
```

Also, the docs at https://docs.saltproject.io/en/latest/ref/modules/all/salt.modules.salt_version.html claim you should also be able to pass a version number, but this results in a traceback:

```
salt salt01 salt_version.less_than '3006'
salt01:
    The minion function caused an exception: Traceback (most recent call last):
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/minion.py", line 1935, in _thread_return
        return_data = minion_instance._execute_job_function(
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/minion.py", line 1894, in _execute_job_function
        return_data = self.executors[fname](opts, data, func, args, kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 149, in __call__
        return self.loader.run(run_func, *args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1232, in run
        return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1247, in _run_as
        return _func_or_method(*args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/executors/direct_call.py", line 10, in execute
        return func(*args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 149, in __call__
        return self.loader.run(run_func, *args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1232, in run
        return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1247, in _run_as
        return _func_or_method(*args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/modules/salt_version.py", line 145, in less_than
        if _check_release_cmp(name) == -1:
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/modules/salt_version.py", line 161, in _check_release_cmp
        map_version = get_release_number(name)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/modules/salt_version.py", line 69, in get_release_number
        name = name.lower()
    AttributeError: 'int' object has no attribute 'lower'

salt salt01 salt_version.less_than '3006.0'
salt01:
    The minion function caused an exception: Traceback (most recent call last):
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/minion.py", line 1935, in _thread_return
        return_data = minion_instance._execute_job_function(
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/minion.py", line 1894, in _execute_job_function
        return_data = self.executors[fname](opts, data, func, args, kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 149, in __call__
        return self.loader.run(run_func, *args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1232, in run
        return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1247, in _run_as
        return _func_or_method(*args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/executors/direct_call.py", line 10, in execute
        return func(*args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 149, in __call__
        return self.loader.run(run_func, *args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1232, in run
        return self._last_context.run(self._run_as, _func_or_method, *args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/loader/lazy.py", line 1247, in _run_as
        return _func_or_method(*args, **kwargs)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/modules/salt_version.py", line 145, in less_than
        if _check_release_cmp(name) == -1:
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/modules/salt_version.py", line 161, in _check_release_cmp
        map_version = get_release_number(name)
      File "/opt/saltstack/salt/lib/python3.10/site-packages/salt/modules/salt_version.py", line 69, in get_release_number
        name = name.lower()
    AttributeError: 'float' object has no attribute 'lower'
```

**Setup**
(Please provide relevant configs and/or SLS files (be sure to remove sensitive info. There is no general set-up of Salt.)

Please be as specific as possible and give set-up details.

- [ ] on-prem machine
- [x] VM (Virtualbox, KVM, etc. please specify)
- [ ] VM running on a cloud service, please be explicit and add details
- [ ] container (Kubernetes, Docker, containerd, etc. please specify)
- [ ] or a combination, please be explicit
- [ ] jails if it is FreeBSD
- [ ] classic packaging
- [x] onedir packaging
- [ ] used bootstrap to install

**Steps to Reproduce the behavior**

```
salt salt01 salt_version.get_release_number 'Sulfur'
salt01:
    No version assigned.

salt salt01 salt_version.equal 'Sulfur'
salt01:
    False

salt salt01 salt_version.less_than 'Sulfur'
salt01:
    True
```

**Expected behavior**

```
salt salt01 salt_version.get_release_number 'Sulfur'
salt01:
    3006

salt salt01 salt_version.equal 'Sulfur'
salt01:
    True

salt salt01 salt_version.less_than 'Sulfur'
salt01:
    False
```

```
salt salt01 salt_version.less_than '3006'
salt01:
    False

salt salt01 salt_version.less_than '3006.0'
salt01:
    False
```

**Versions Report**
<details><summary>salt --versions-report</summary>
(Provided by running salt --versions-report. Please also mention any differences in master/minion versions.)

```yaml
salt --versions-report
Salt Version:
              Salt: 3006.1

Python Version:
            Python: 3.10.11 (main, May  5 2023, 02:31:54) [GCC 11.2.0]

Dependency Versions:
              cffi: 1.14.6
          cherrypy: unknown
          dateutil: 2.8.1
         docker-py: Not Installed
             gitdb: Not Installed
         gitpython: Not Installed
            Jinja2: 3.1.2
           libgit2: 1.5.0
      looseversion: 1.0.2
          M2Crypto: Not Installed
              Mako: Not Installed
           msgpack: 1.0.2
      msgpack-pure: Not Installed
      mysql-python: Not Installed
         packaging: 22.0
         pycparser: 2.21
          pycrypto: Not Installed
      pycryptodome: 3.9.8
            pygit2: 1.11.1
      python-gnupg: 0.4.8
            PyYAML: 5.4.1
             PyZMQ: 23.2.0
            relenv: 0.12.3
             smmap: Not Installed
           timelib: 0.2.4
           Tornado: 4.5.3
               ZMQ: 4.3.4

Salt Extensions:
 saltext.bitwarden: 0.0.1b14

System Versions:
              dist: almalinux 8.7 Stone Smilodon
            locale: utf-8
           machine: x86_64
           release: 4.18.0-425.19.2.el8_7.x86_64
            system: Linux
           version: AlmaLinux 8.7 Stone Smilodon
```
</details>

**Additional context**

### PR DESCRIPTION
**Pull request by meaksh**
### What does this PR do?

This PR fixes an issue and some kind of mess we have around Salt version detection and codename resolution.

Currently, the codename resolution (`salt_version.get_release_number` and `salt_version.equal`) seems broken in Salt:

```console
# pip3 install salt
Collecting salt
  Downloading salt-3006.1.tar.gz (20.4 MB)
[...]
Successfully installed Jinja2-3.0.3 MarkupSafe-2.0.1 PyYAML-6.0 certifi-2023.5.7 charset-normalizer-2.0.12 contextvars-2.4 distro-1.8.0 idna-3.4 immutables-0.19 jmespath-0.10.0 looseversion-1.2.0 msgpack-1.0.5 psutil-5.9.5 pycryptodomex-3.18.0 pyzmq-25.0.2 requests-2.27.1 salt-3006.1 typing-extensions-4.1.1 urllib3-1.26.16

# salt-call --version
salt-call 3006.1

# salt-call --local grains.get saltversion
local:
    3006.1

# salt-call --local salt_version.equal Sulfur
local:
    False

# salt-call --local salt_version.get_release_number Sulfur
local:
    No version assigned.
```

### Previous Behavior
As mentioned above, the `salt_version.get_release_number` and `salt_version.equal` execution module functions (and probably others) are not working fine.

Other codenames seems working fine:

```console
# salt-call --local salt_version.get_release_number Neon
local:
    3000
```

After some debugging, I've noticed the following situations:
- Codename definitions for Salt >= 3006 seems wrong at `salt/version.py`, as they map the version codename to the `300X.0` release but not to the entire `300X`. This causes problems on codename detection.
- There seems to be a mess with the usage of `can_have_dot_zero` function that causes wrong a detection of Salt versions.

### New Behavior
Salt version resolution works fine now and versions are detected as expected:

```console
# salt-call --local salt_version.get_release_number Sulfur
local:
    3006

# salt-call --local salt_version.equal Sulfur
local:
    True
```

### Merge requirements satisfied?
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: tests/pytests/unit/modules/test_salt_version.py [hunk 1]
```diff
@@ -21,7 +21,7 @@ def test_mocked_objects():
     for k, v in salt.version.SaltStackVersion.LNAMES.items():
         assert k == k.lower()
         assert isinstance(v, tuple)
-        if sv.new_version(major=v[0]) and not sv.can_have_dot_zero(major=v[0]):
+        if sv.new_version(major=v[0]):
```
Comment (s0undt3ch) [inline tests/pytests/unit/modules/test_salt_version.py:24]: So, from 3006 onward, the `.0` must always be present, the existing logic was correct, but SaltStackVersion is now reporting it differently?

File: tests/pytests/unit/modules/test_salt_version.py [hunk 2]
```diff
@@ -64,6 +64,13 @@ def test_get_release_number_success_new_version():
     assert salt_version.get_release_number("Neon") == "3000"
 
 
+def test_get_release_number_success_new_version_with_dot():
+    """
+    Test that a version is returned for new versioning (3006)
+    """
+    assert salt_version.get_release_number("Sulfur") == "3006"
```
Comment (s0undt3ch) [inline tests/pytests/unit/modules/test_salt_version.py:71]:
```suggestion
    assert salt_version.get_release_number("Sulfur") == "3006.0"
```
In theory, this should be the expected outcome... Although, all 3006 minor versions are Sulfur versions...

File: tests/pytests/unit/modules/test_salt_version.py (hunk 1 above)
Comment (s0undt3ch) [reply tests/pytests/unit/modules/test_salt_version.py:24]: Ok, I see why this had to change.
And it looks like it hasn't broken other functionality.
```
❯ python
Python 3.10.2 (main, Nov 15 2022, 11:41:37) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import salt.version
>>> v = salt.version.SaltStackVersion.parse("3006")
>>> str(v)
'3006.0'
>>> v.info
v.info
>>> v.info
(3006, 0)
>>> v = salt.version.SaltStackVersion.parse("3005")
>>> str(v)
'3005'
>>> v.info
(3005,)
>>>
```

**Review by s0undt3ch — APPROVED** [review]

File: changelog/64554.fixed.md
Comment (s0undt3ch) [inline changelog/64554.fixed.md:1]: This file should be named `64306.fixed.md`

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

File: changelog/64554.fixed.md
Comment (meaksh) [reply changelog/64554.fixed.md:1]: Done. Thanks for the review!

**Comment by s0undt3ch** [comment]
@meaksh since you've fixed most of the issues, could you also add a test case for the error shown in https://github.com/saltstack/salt/issues/64306 ?
Let me know if you don't have the bandwidth and I can push it to your PR.

**Comment by meaksh** [comment]
@s0undt3ch ok, I'll give it a try to implement a test that covers the error shown at https://github.com/saltstack/salt/issues/64306

**Comment by meaksh** [comment]
@s0undt3ch I've added the mentioned test.

I realized `salt_version.less_than` and others are expecting to get a "codename" string as input, and not really a version number (as it was done in https://github.com/saltstack/salt/issues/64306).

I've just adjusted to the code to raise a `CommandExecutionError` instead.

**Comment by s0undt3ch** [comment]
> @s0undt3ch I've added the mentioned test.
> 
> I realized `salt_version.less_than` and others are expecting to get a "codename" string as input, and not really a version number (as it was done in #64306).
> 
> I've just adjusted to the code to raise a `CommandExecutionError` instead.

Thanks!

**Review by s0undt3ch — DISMISSED** [review]

**Review by Ch3LL — CHANGES_REQUESTED** [review]
The documentation for `get_release_number` states it should return `MAJOR.PATCH` but your PR is only returning `MAJOR` for 3006 which now does include a patch in the version. The expected behavior seems to be to return `3006.0` instead of `3006`.

**Review by MKLeb — DISMISSED** [review]

**Comment by ggiesen** [comment]
If the intention is to not 

> > @s0undt3ch I've added the mentioned test.
> > I realized `salt_version.less_than` and others are expecting to get a "codename" string as input, and not really a version number (as it was done in #64306).
> > I've just adjusted to the code to raise a `CommandExecutionError` instead.
> 
> Thanks!

If the intention is not to fix this, then please update the example in the docs:

https://github.com/saltstack/salt/blob/9aeed74e96542558fc473a24d6b38add4a62b65b/salt/modules/salt_version.py#L18-L29

**Comment by meaksh** [comment]
@Ch3LL as mentioned, the documentation for `salt_version` is now upgraded to reflect the expected behavior.

**Review by s0undt3ch — APPROVED** [review]

**Comment by s0undt3ch** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|3006.x|[](https://github.com/saltstack/salt/pull/64671)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Comment by s0undt3ch** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|3006.x|[](https://github.com/saltstack/salt/pull/64671)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Comment by s0undt3ch** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|3006.x|[](https://github.com/saltstack/salt/pull/64671)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Comment by s0undt3ch** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|3006.x|[](https://github.com/saltstack/salt/pull/64671)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Comment by s0undt3ch** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|3006.x|[](https://github.com/saltstack/salt/pull/64671)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Comment by s0undt3ch** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|3006.x|[](https://github.com/saltstack/salt/pull/64671)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Comment by s0undt3ch** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|3006.x|[](https://github.com/saltstack/salt/pull/64671)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Comment by s0undt3ch** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|3006.x|[](https://github.com/saltstack/salt/pull/64671)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Comment by s0undt3ch** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|3006.x|[](https://github.com/saltstack/salt/pull/64671)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Comment by s0undt3ch** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|3006.x|[](https://github.com/saltstack/salt/pull/64671)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

**Pull request merged**